dev/archery/archery/linking.py

```diff
@@ -61,9 +64,71 @@ def list_dependency_names(self):
             names.append(name)
         return names
 
+    def _extract_symbols(self, symbol_info):
+        return [re.search(r'\S+$', line).group() for line in symbol_info if line]
+
+    def _remove_weak_symbols(self, lines):
+        return [line for line in lines if not re.search(r'\s[Ww]\s', line)]
+
+    def list_symbols_for_dependency(self, dependency, remove_symbol_versions=False):
+        result = _nm.run('-D', dependency, stdout=subprocess.PIPE)
+        lines = result.stdout.decode('utf-8').splitlines()
+        if remove_symbol_versions:
+            lines = [re.split('@@', line)[0] for line in lines]
+        return self._extract_symbols(lines)
+
+    def list_undefined_symbols_for_dependency(self, dependency,
+                                              remove_symbol_versions=False):
+        result = _nm.run('-u', dependency, stdout=subprocess.PIPE)
+        lines = result.stdout.decode('utf-8').splitlines()
+        if remove_symbol_versions:
+            lines = [re.split('@@', line)[0] for line in lines]
+        lines = self._remove_weak_symbols(lines)
+        return self._extract_symbols(lines)
+
+    def find_library_paths(self, libraries):
+        paths = {}
+        system = platform.system()
+        for lib in libraries:
+            paths[lib] = []
+            if system == 'Linux':
+                result = _ldconfig.run('-p', stdout=subprocess.PIPE)
```

Comment:
Why do you use `ldconfig` not `ldd`?
In general, `ldconfig` exists at `/sbin/ldconfig` not `/bin/` and normal user's `PATH` doesn't include `/sbin/`. So `ldd` is better than `ldconfig`.

Reply:
I am getting a missing symbol after using `ldd` locally, but I am going to add this here and see what happens in the CIs

```bash
archery linking check-dependencies \
  --allow ld-linux-aarch64 \
  --allow ld-linux-x86-64 \
  --allow libc \
  --allow libdl \
  --allow libgcc_s \
  --allow libm \
  --allow libpthread \
  --allow librt \
  --allow libstdc++ \
  --allow libz \
  --allow linux-vdso \
  /home/vibhatha/Documents/Work/Apache_Arrow/issues/gh-40018/arrow-dataset-15.0.2/x86_64/libarrow_dataset_jni.so
Error: Undefined symbols found in /home/vibhatha/Documents/Work/Apache_Arrow/issues/gh-40018/arrow-dataset-15.0.2/x86_64/libarrow_dataset_jni.so:
__tls_get_addr

```

Reply:
It seems that `__tls_get_addr` is included in `/lib64/ld-linux-x86-64.so.2`:

```console
$ nm -C /lib64/ld-linux-x86-64.so.2  | grep __tls_get_addr
0000000000012c40 t __GI___tls_get_addr
0000000000012c40 t ___tls_get_addr
00000000000179c0 T __tls_get_addr
0000000000012db0 t __tls_get_addr_slow
```

Reply:
Yes so when we do the `ldd` we cannot find that library unfortunately because it is not in the `allowed` list, right?

Reply:
`ld-linux-x86-64` is included in `allowed` list. See the command line you showed in https://github.com/apache/arrow/pull/40520#discussion_r1555116608 .

Reply:
Oh I see what is going on here

```bash
	linux-vdso.so.1 (0x00007ffef15e1000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x0000711a06894000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x0000711a0688f000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x0000711a0688a000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x0000711a03000000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x0000711a03319000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x0000711a032f9000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x0000711a02c00000)
	/lib64/ld-linux-x86-64.so.2 (0x0000711a068ad000)
```
My filtering logic misses the last one.

Reply:
Also `linux-vdso.so.1` for this I don't get a path, and `nm` won't work on this. Any thought about this?

Reply:
We can ignore it.

Reply:
This is virtual one? What's actually happening here? My knowledge is limited here?

Reply:
It's a virtual shared library. So it doesn't exist in file system. (We can't get symbols of it.)

Reply:
Got it. Thanks @kou